karapace/kafka_rest_apis/__init__.py

```diff
@@ -51,16 +51,48 @@ def __init__(self, config: Config) -> None:
         self._add_kafka_rest_routes()
         self.serializer = SchemaRegistrySerializer(config=config)
         self.proxies: Dict[str, "UserRestProxy"] = {}
+        self._proxy_lock = asyncio.Lock()
         log.info("REST proxy starting with (delegated authorization=%s)", self.config.get("rest_authorization", False))
+        self._idle_proxy_janitor_task: Optional[asyncio.Task] = None
 
     async def close(self) -> None:
+        if self._idle_proxy_janitor_task is not None:
+            self._idle_proxy_janitor_task.cancel()
+            self._idle_proxy_janitor_task = None
         async with AsyncExitStack() as stack:
             stack.push_async_callback(super().close)
             stack.push_async_callback(self.serializer.close)
 
             for proxy in self.proxies.values():
                 stack.push_async_callback(proxy.aclose)
 
+    async def _idle_proxy_janitor(self) -> None:
+        IDLE_PROXY_TIMEOUT = 5 * 60
+        while True:
+            await asyncio.sleep(IDLE_PROXY_TIMEOUT / 2)
+
+            key, proxy = None, None
+            async with self._proxy_lock:
+                # Always clean one at time per iteration, don't mutate dict while iterating
+                for _key, _proxy in self.proxies.items():
+                    # If UserRestProxy has consumers with state, disconnecting loses state
+                    if _proxy.num_consumers() > 0:
+                        timeout = self.config.get("consumer_idle_disconnect_timeout", 0)
```

Comment:
Would be nice to have a tests.

Reply:
To test this with integration tests would require setting up Kafka with authorization. Most of the new code in this PR is not unit-testable, thus test coverage is quite low. To test this specific feature with the integration test, the test would have to wait up to minutes as this task is not so frequently checking for idle proxies, ie. 2.5 minutes per cycle. I'm kind out of tangible ideas how to really test this.